test/integration/scheduler_perf/config/performance-config.yaml

```diff
@@ -688,48 +688,103 @@
 #      initPods: 20000
 #      measurePods: 5000
 
-- name: Unschedulable
+# Measure the time it takes to schedule default measurePods after scheduling 20% of high priority Pods that trigger preemption.
```

Comment:
What's the core goal of this test case? 
If it aims to capture the latency of scheduling that includes preemption, then I don't understand why we need normal Pods created at the last `createPods` in the first place. They just blur the results from the Pods that have triggered the preemption.

Reply:
The goal is to measure how long a scheduling of lower pririty pods has to wait when there is a preemption process in front of it.

IIUC the overall time for preemption measured in PreemptionBasic test consist of both sync and async part (obviously there is no async part yet). This test is meant to measure only the sync part (preeption sync + measure sync) ignoring the async part, as preeption step has `skipWaitToCompletion: true`

Reply:
So... sorry, but still don't quite understand. 
So, you're trying to measure the scheduling throughput for other Pods, which are completely unrelated to the preemption, right? Why do we want to measure it? How could the scheduling for those Pods be affected by the Pods triggering preemptions?

Reply:
No problem. I might be misinterpreting something, but the idea is that the time of the preemption and scheduling of these unrelated Pods is measured together. Note that the test should not wait for the preemption pods to finish the asynchronous part (assigning nodes to pods) because of `skipWaitToCompletion: true`, so the sooner the preemption process finishes its synchronous part, the shorter this time will be:

1. Schedule low priority initial pods and wait for completion (pods have nodes assigned)
2. Start time measure
3. Schedule high priority pods, but don't wait for completion
4. Schedule regular pods and wait for completion (pods have nodes assigned)
5. Stop time measure (preemption pods might still now have nodes assigned)

The main goal is to measure the time the regular pods are waiting for the scheduler, so in fact, show the improvement brought by the asynchronous preemption feature in the scheduler.

If there was no `skipWaitToCompletion: true`, then sure, the test would always measure the time of the preemption including node assignment (the asynchronous part) and scheduling of regular pods, so would not show the improvement.

Reply:
I see, so we want to make sure that pods that are created right after pods that need preemption can be scheduled in a timely fashion.

That's different from `PreemptionBasic`. So we need a bit of clarification in the comments. I would start the comment with "what are we testing" and follow with "how are we achieving the scenario" (which priorities or sizes are at play).

Reply:
> By waiting for completion (binding) of the regular nodes, we know we included in our measuments only the synchronous part of the preemption process.

So, does `measurePods` exist only for knowing the end of all scheduling cycles of `blockingPods`? like

1. The scheduler starts to schedule `blockingPods`, which should result in the preemption.
2. After (1), the scheduler starts to schedule `measurePods`. We don't really care about them because the preemption doesn't get involved in their scheduling. But, we need it so that we can know the end of all scheduling cycles for `blockingPods`, that should trigger the preemption.

Am I right?
Then, why do we need to create such a large number of `measurePods`? If my hypothesis is correct, even 1 Pod is enough in `measurePods`, no?

Reply:
@alculquicondor 

> what completes the preemption of the pods?

When all Pod API calls for deleting all victim Pods + updating the Pod with `nominatedNodeName` is complete, then the preemption goroutine is complete.

Reply:
> Then, why do we need to create such a large number of `measurePods`? If my hypothesis is correct, even 1 Pod is enough in `measurePods`, no?

That seems true. But measuring just one pod might make the result noisy. But I can't think of a better alternative.

> When all Pod API calls for deleting all victim Pods + updating the Pod with `nominatedNodeName` is complete, then the preemption goroutine is complete.

Sorry, I meant the whole preemption, not the routine. What terminates the existing pods so that the preemptor pods can schedule?

Reply:
>  But I can't think of a better alternative.

Maybe a way would be to create the preemptor pods and "measure" pods in an interleaved manner. Or create preemptor pods in a routine, while creating the measure pods.

But if it's starting to sound too complicated, maybe `PreemptionBasic` is enough to show any improvement in the throughput, even if it doesn't target specifically the change to make preemption asynchronous.

Reply:
The way this test is written was supposed to measure the time the regular pods needs to wait for scheduling.

The problem is that we need to express it in the form of throughput, which depends not only on the time the regular pods wait, but also on the number of them. So the higher number of regular pods there are in the test, the better throughput we get, but at the same time the higher number of preemption pods there are, the througput drops. This is why the test has the fixed ratio, however the value is debatable.

However, when I looked deeper how the throuput is measured, I discovered a few problems:
1. The time starts counting when the first pod got scheduled (node is got assigned).
2. Since we don't wait for preemption to finish, we don't know how many of the preemption pods managed to schedule (thus how they impact the final throuput).

One approach (A) would be to indeed measure only the throughput of the regular nodes including the waiting time, but becase of 1., the time starts once scheduler finishes preemptions (without binding), so it in fact shows the pure throuput of regular pods. We theoretically could change the 1. logic and start the timer earlier, but I guess this logic is there for some reason.

Another one (B) is setting additional barier to wait for the remaining preemption pods to be binded. We would get more consistent results, but obviously we would get throuput of both sync and async parts, so the time improvement would not be visible.

Approach (A) is cleaner, but colides with the current implementation. I think we can look into that. I discovered a few issues with the current metric collection implementation and will create an issue for that.

Reply:
Maybe before I create an issue, let me enumerate them here, as it's very likely that the collection work this way in purpose:

1. Time collection starts time measurement when the first pod got scheduled instead of when the collection operation  (`startCollectingMetrics`) begins.
2. Moreover, the time start is aligned to a second that opens the collection window (every 1 second ticks), therefore measures of pods scheduled in the first second are ignored (scheduling operations below 1s may go unnoticed)
3. Pods scheduled in the last window are not measured either, which often leads to not collecting any throughput data for tests that last below 2 seconds.

In effect, thoughput is calculated (if any) based on completed time windows, but first and last second is ignored, which often may lead to getting incorrect results. I think that calculating average thoughput (and percentiles) precisely between start and stop should be also possible.

Reply:
Answering to

> Maybe before I create an issue, let me enumerate them here, as it's very likely that the collection work this way in purpose:

1. Assuming the number of pods to schedule is not minuscule, starting the measuring earlier might not be that significant of a change. But yes, starting it when `startCollectingMetrics` begins seems more appropriate.
2. I think this might be related to the fact that API timestamps have a second granularity. But I would welcome any opportunity for improvement.
3. I think that one was intentional, as there might not be enough pods in the last second, polluting the "throughput"

> The way this test is written was supposed to measure the time the regular pods needs to wait for scheduling.

I'm not too sure that we should be doing that. Just measuring throughput of pods that need preemption is sort of equivalent, because all pods need preemption, so they are all waiting for others to finish.
Moreover, we are trying to focus on throughput, not latency.

Reply:
> I'm not too sure that we should be doing that. Just measuring throughput of pods that need preemption is sort of equivalent, because all pods need preemption, so they are all waiting for others to finish.
Moreover, we are trying to focus on throughput, not latency.

Measuring a throughput of the preemption itself after we introduce async feature will become also problematic, as due to the way it's currently measured (time starts counting when the first pod is bind) we would calculate only the throughput of the async part (in case the sync part is already finished before the first pod is bind).

We can actually stick to the approach with mixing different pods and schedule 50% of regular pods and 50% of pods that require preemption. However, we'd schedule half of the regular ones first (25%), all preemption pods (50%) and the remaining half (25%). It will guarantee we start measure before the preemption process starts and give it some time to finish (enclose it). The only downside is that the throughput of the preemption will have approximately 50% contribution to the result. On the other hand, the throughput of the regular nodes is more or less known.

We would not have to change the collection logic which works best for the pods in the middle only WDYT?

Reply:
> Measuring a throughput of the preemption itself after we introduce async feature will become also problematic, as due to the way it's currently measured (time starts counting when the first pod is bind) we would calculate only the throughput of the async part (in case the sync part is already finished before the first pod is bind).

I agree, but, with a sufficient amount of pods, the measurement will be roughly equivalent.

Regardless, changing the collection logic seems to be a reasonable move. And we can start thinking of the idea of mixing or interleaving pods with different templates, as opposed to what we have today: create all pods of one template, then another template.

Reply:
> I agree, but, with a sufficient amount of pods, the measurement will be roughly equivalent.

I also agree, but the point is to see how regular pods interleave with the asynchronous part of the preemption process (make use of scheduler time while asynchronous part is in progress). So before the async preemption is implemented, both processes will be sequential, but later, the asynchronous part will be executed in parallel to scheduling.

> Regardless, changing the collection logic seems to be a reasonable move. 

I think we need to either change the logic (start time calculation before preemption starts) or schedule a few regular pods in front of it to kick off time measurement. I'm fine with both.

> And we can start thinking of the idea of mixing or interleaving pods with different templates, as opposed to what we have today: create all pods of one template, then another template.

What interleaving pods would give over scheduling them one after another? The async part would be interleaved (exactly what we expect to happen) and the sync part would be sequential anyway. Sure, the interleaved scheduling would give more predictable results (better control over the fraction of preemption vs regular pods), but the question is whether it would make a real difference. I guess we could achieve similar behavior by scheduling both pods at the same priority and let the queue process them in random order.

Reply:
I think I understand why the test approach I proposed is confusing. I believe the performance test author was expecting the tests to produce stable scheduling throughput and then measure it. Ignoring the fist and the last second would be completely fine.

The test approach I proposed is creating a more complex scenario assuming we are able to measure the throughput more precisely between start and end of this test scenario. So instead of changing the current metric collection approach, we could indeed implement a test in the way that it interleaves preemption pods with regular pods in a loop creating more complex, but repeatable flow.

Reply:
I rewrote the test to use churn operation, that will inject high priority preemption pod every 50 milliseconds (20 per second). It allowed to reduce the measured pods throughput from approx 1000 to 300, so we should be able to notice the difference with async preemption.

For simplicity, I measure throughput of only regular pods, as the preemption ones would contribute only in small fraction to that. It is indeed simpler to interpret.